Tiltfile

```diff
@@ -61,11 +62,89 @@ def deploy_capi():
+    """
+else:
+    tilt_dockerfile_header_cso = """
+    FROM ghcr.io/sovereigncloudstack/cso:v0.1.0-alpha.2 as builder
+
+    FROM docker.io/library/alpine:3.18.0 as tilt
+    WORKDIR /
+    COPY --from=builder /usr/local/bin/helm /usr/local/bin/helm
+    COPY --from=builder /manager /manager
+    """
+    
+
 def deploy_cso():
-    version = settings.get("cso_version")
-    cso_uri = "https://github.com/sovereignCloudStack/cluster-stack-operator/releases/download/{}/cso-infrastructure-components.yaml".format(version)
-    cmd = "curl -sSL {} | {} | kubectl apply -f -".format(cso_uri, envsubst_cmd)
-    local(cmd, quiet = True)
+    # yaml = str(kustomizesub("./hack/observability")) # build an observable kind deployment by default
+    yaml_cso = './cso-components.yaml'
+    local_resource(
+        name = "cso-components",
+        cmd = ["sh", "-ec", sed_cmd, yaml_cso, "|", envsubst_cmd],
+        labels = ["CSO"],
+    )
+
+    entrypoint = ["/manager"]
+    extra_args = settings.get("extra_args")
+    if extra_args:
+        entrypoint.extend(extra_args)
+
+    # Set up an image build for the provider. The live update configuration syncs the output from the local_resource
+    # build into the container.
+    if settings.get("local_mode"):
+        docker_build_with_restart(
+            ref = "ghcr.io/sovereigncloudstack/cso-staging",
+            context = ".",
+            dockerfile_contents = tilt_dockerfile_header_cso,
+            target = "tilt",
+            entrypoint = entrypoint,
+            live_update = [
+                sync(".release", "/tmp/downloads/cluster-stacks"),
+            ],
+            ignore = ["templates"],
+        )
+    else:
+        docker_build_with_restart(
+            ref = "ghcr.io/sovereigncloudstack/cso-staging",
+            context = ".",
+            dockerfile_contents = tilt_dockerfile_header_cso,
+            target = "tilt",
+            entrypoint = entrypoint,
+            live_update = [
+                sync("cso-components.yaml", "/cso-components.yaml"),
+            ],
+            ignore = ["templates"],
+        )
+
+    k8s_yaml(yaml_cso)
+    k8s_resource(workload = "cso-controller-manager", labels = ["CSO"])
+    k8s_resource(
+        objects = [
+            "cso-system:namespace",
+            "clusteraddons.clusterstack.x-k8s.io:customresourcedefinition",
+            "clusterstackreleases.clusterstack.x-k8s.io:customresourcedefinition",
+            "clusterstacks.clusterstack.x-k8s.io:customresourcedefinition",
+            "cso-controller-manager:serviceaccount",
+            "cso-leader-election-role:role",
+            "cso-manager-role:clusterrole",
+            "cso-leader-election-rolebinding:rolebinding",
+            "cso-manager-rolebinding:clusterrolebinding",
+            "cso-serving-cert:certificate",
+            "cso-selfsigned-issuer:issuer",
+            "cso-validating-webhook-configuration:validatingwebhookconfiguration",
```

Comment:
This line should probably be commented out, otherwise I got an error: `Internal error occurred: failed calling webhook "validation.clusterstack.clusterstack.x-k8s.io"`

Reply:
CSO has commented it here https://github.com/SovereignCloudStack/cluster-stack-operator/blob/main/Tiltfile#L210 but probably here it somehow should work because I do see cso-validating-webhook-configuration in the cso-components.yaml.
But maybe we won't need that according to my previous comments, e.g. when cso-components.yaml will be removed from local and downloaded from CSO release instead

Reply:
It doesn't work even when this line is commented.

Reply:
I have to comment out also ValidatingWebhookConfiguration from cso-components.yaml but then GIT_ env vars are missing:
```bash
$ k get clusterstack -n cluster 
NAME           PROVIDER    CLUSTERSTACK   K8S    CHANNEL   AUTOSUBSCRIBE   USABLE   LATEST   AGE   REASON                        MESSAGE
clusterstack   openstack   scs            1.27   stable    false                             99s   GitTokenOrEnvVariableNotSet   failed to create git config: environment variable GIT_PROVIDER is not set
```
It is because secret cso-cluster-stack-variables, as I pointed out in https://github.com/SovereignCloudStack/cluster-stack-provider-openstack/pull/78#discussion_r1484114317, is missing and also there is no env from secret mapping in the cso deployment.

Reply:
> Actually, it seems like there is no webhook server running in cso manager, am I right? 

Yes, there is not webhook server running in cso as of now but it's a WIP from my side and I mixed it up. I just use `kustomize build` against my development branch and got the `cso-components.yaml` file. I'll sync/clean the cso-components file with what we have in GitHub release.